sonar-core/src/main/java/org/sonar/core/component/Module.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.core.component;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+import org.sonar.core.platform.ComponentContainer;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public abstract class Module {
+  private ComponentContainer container;
+
+  public Module configure(ComponentContainer container) {
+    this.container = checkNotNull(container);
+
+    configureModule();
+
+    return this;
+  }
+
+  protected abstract void configureModule();
+
+  protected void add(@Nullable Object object, boolean singleton) {
+    if (object != null) {
+      container.addComponent(object, singleton);
+    }
+  }
+
+  protected <T> T getComponentByType(Class<T> tClass) {
+    return container.getComponentByType(tClass);
+  }
+
+  protected void add(Object... objects) {
```

Comment:
`add(Object, boolean)` is currently never used. It created it only to provide support for the current use of `ComponentContainer.add(Object, boolean)` in `PlatformLevel`. Btw, the `ComponentContainer` method is used only once with a false parameter and I don't think it is even appropriate https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/platform/platformlevel/PlatformLevel4.java#L424.

I'm not very happy with the idea of `protected abstract void configureModule(ComponentContainer container)`.

First, it creates a non necessary coupling between subclasses of `Module` and the class `ComponentContainer`. Implementation only cares to add classes or objects to the module. How this objects are in turn added to the container is implementation specific. Side-benefit: if implementation is someday not based on `ComponentContainer` but, eg. on PicoContainer directly, Module subclasses will not change.

Second, `ComponentContainer` exposes to many methods to add a component to the container. You note that methods `add` of module are confusing, this is worse for `ComponentContainer` imo (look at all the add* methods). In addition, some methods have nothing to do with adding a component and they should not be called from a module.